docs/guides/extending-beluga.md

```diff
@@ -1,3 +1,264 @@
+
+This model computes the likelihood of a measurement given the particle's state and the environment. It also accounts for the uncertainty in the measurements.
+
+### Implementation
+
+In this library, sensor models are implemented as callable objects that take a measurement and return another callable, a state weighting function.
+This function can be applied to each particle to compute its weight or likelihood based on its current state.
+
+This is how the sensor model is used in the particle filter algorithm shown above:
+
+```cpp
+particles |= beluga::actions::reweight(sensor_model(measurement));
+```
+
+This functional composition can be explained as follows:
+
+1. The sensor model takes the measurement as input to generate a function that dictates how each particle’s weight should be adjusted. It holds an internal reference to a map of the environment to correctly determine the likelihood of each state.
+2. The `reweight` function takes the state weighting function and creates a range action that can be applied to a range of particles.
+3. The `|=` applies the range action to the entire set of particles, updating the weights of all particles according to the model.
+
+This is the basic template we will use for creating a custom sensor model:
+
+```cpp
+struct SensorModel {
+ public:
+  using state_type = ...;
+  using weight_type = ...;
+  using measurement_type = ...;
+
+  auto operator()(measurement_type measurement) const {
+    ...
+    return [=](state_type state) -> weight_type { ... };
+  }
+};
+```
+
+Let’s implement a simple sensor model of an agent moving across a 1-dimensional world. For this we will use the following definitions:
+
+- The state will be the position of the agent along an infinite line. Represented as a `double`.
+- The weight type will be `double`.
+- The sensor modeled is a landmark detector, capable of measuring the distance between the agent and nearby landmarks.
+- Measurements consist of an array of distances from the agent to each visible landmark.
+- All landmarks are indistinguishable from one another, treated as identical in the context of measurement.
+- The likelihood of each measurement is calculated using a Gaussian function centered at the distance to the nearest landmark.
+- The weight of each particle is calculated by aggregating the likelihoods derived from all detections, transformed to the particle's frame of reference.
+- The model's parameters include the standard deviation of the Gaussian distribution, which models the detection noise, and a map listing the positions of all landmarks.
+
+The resulting implementation could look like this:
+
+```cpp
+struct SensorModel {
+ public:
+  using state_type = double;
+  using weight_type = double;
+  using measurement_type = std::vector<double>;
+
+  SensorModel(double sigma, std::vector<double> map) : sigma_{sigma}, map_{std::move(map)} {}
```

Comment:
nit: just to follow the same style as the motion model struct add the `explicit` to the constructor